ahocorasick.go

```diff
@@ -232,16 +235,14 @@ func (m *Matcher) Match(in []byte) []int {
 			f := n.child[c]
 			n = f
 
-			if f.output && f.counter != m.counter {
+			if f.output && !found(f.index, hits) {
```

Comment:
Instead of linear lookup in `hits`, we can create a slice of booleans local to the function (`hits`) and just use `node.index` as index in it.

Reply:
Further optimization would be to have a pool of those slices of booleans to avoid allocating them on every `Match()` call. Or even take that slice as an argument, then allocating will be the problem of the caller.

Reply:
Even better, you can reinstate `Matcher.counter` to count the "generations" and keep that generation inside that new slice (instead of booleans). Then you don't have to reset the slice to all `false` values every time. Basically, the same logic as was originally, but we store the generation number in a separate place instead of directly inside the nodes.

Reply:
Thanks for your suggestions, **kvap**!

However, those changes will introduce extra allocation per each search, and, to be honest, will make code more complex. In order to probably reduce allocations, we might end up introducing another `sync.pool` to manage a bounded amount of those slices.

That is not what I personally want to do here. I agree that this simple, but a dumb method of iterating through hits will cost more CPU when a number of hits would be significantly large, looping through hits with every new hit being found would cost more and more.

I decided to rewrite it as a bit search. Instead of introducing another slice or map and manage it, I think we can use here an integer set to 0. And set a bit (`f.index`) every time we put it in the `hits`. Then we can easily check that bit and decide whether we should store it or not.

The `hits` is a slice of ints, which supposed to be unique. So the max possible value is limited by `int`, therefore we can use our mask as `int`.

Reply:
This obviously won't work on indices > 63/31 (depends on the system). Will rework it.